src/Compatibility/Microsoft.DotNet.ApiCompatibility/ValidationSuppressions/SuppressionEngine.cs

```diff
@@ -0,0 +1,208 @@
+
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Xml.Serialization;
+
+namespace Microsoft.DotNet.ValidationSuppression
+{
+    /// <summary>
+    /// Collection of Suppressions which is able to add suppressions, check if a specific error is suppressed, and write all suppressions
+    /// down to a file. The engine is thread-safe.
+    /// </summary>
+    public class SuppressionEngine
+    {
+        protected HashSet<Suppression> _validationSuppressions;
+        private ReaderWriterLockSlim _readerWriterLock = new();
+
+        protected SuppressionEngine(string validationSuppressionFile)
+        {
+            _validationSuppressions = ParseValidationSuppressionBaseline(validationSuppressionFile);
+        }
+
+        protected SuppressionEngine()
+        {
+            _validationSuppressions = new HashSet<Suppression>();
+        }
+
+        /// <summary>
+        /// Checks if the passed in error is suppressed or not.
+        /// </summary>
+        /// <param name="diagnosticId">The diagnostic ID of the error to check.</param>
+        /// <param name="target">The target of where the <paramref name="diagnosticId"/> should be applied.</param>
+        /// <param name="left">Optional. The left operand in a APICompat error.</param>
+        /// <param name="right">Optional. The right operand in a APICompat error.</param>
+        /// <returns><see langword="true"/> if the error is already suppressed. <see langword="false"/> otherwise.</returns>
+        public bool IsErrorSuppressed(string? diagnosticId, string? target, string? left = null, string? right = null)
+        {
+            var suppressionToCheck = new Suppression()
+            {
+                DiagnosticId = diagnosticId,
+                Target = target,
+                Left = left,
+                Right = right
+            };
+            return IsErrorSuppressed(suppressionToCheck);
+        }
+
+        /// <summary>
+        /// Checks if the passed in error is suppressed or not.
+        /// </summary>
+        /// <param name="error">The <see cref="Suppression"/> error to check.</param>
+        /// <returns><see langword="true"/> if the error is already suppressed. <see langword="false"/> otherwise.</returns>
+        public bool IsErrorSuppressed(Suppression error)
+        {
+            _readerWriterLock.EnterReadLock();
+            try
+            {
+                if(_validationSuppressions.Contains(error))
+                {
+                    return true;
+                }
+                else
+                {
+                    // See if the error is globally suppressed by checking if the same diagnosticid and target are entered
+                    // without any left and right.
+                    if ((error.DiagnosticId == null || !error.DiagnosticId.StartsWith("pkv", StringComparison.InvariantCultureIgnoreCase)) 
```

Comment:
Could we instead check for API Compat diagnostic prefix? Since that is logic tight to API Compat, if we added more diagnostic IDs later with a different prefix, we've covered those as well.

Reply:
Makes sense, will do.